README.md
# jpeg-parser-hs
A JPEG parser in Haskell for kicks

## Useful links
* [JPEG File Layout and Format](http://vip.sugovica.hu/Sardi/kepnezo/JPEG%20File%20Layout%20and%20Format.htm)
* [JPEG Huffman Coding Tutorial](https://www.impulseadventure.com/photo/jpeg-huffman-coding.html)
* [Unraveling the JPEG](https://parametric.press/issue-01/unraveling-the-jpeg/)
* [Real World Haskell PDF](http://pv.bstu.ru/flp/RealWorldHaskell.pdf)


ChangeLog.md
# Revision history for jpeg-parser

## 0.1.0.0  -- YYYY-mm-dd

* First version. Released on an unsuspecting world.


Parse.hs
module Parse ( Parser
             , Peeler
             , Expecter
             , parser
             , peeler
             , expecter
             , peelAny
             , peelSome
             , peelNBytes
             , expectBytes
             , peelWord8
             , peelWord16
             , peelWord32
             , peelWord64
             , assertThat
             , msnd , mfst
             ) where

import qualified Data.ByteString.Lazy as L
import Data.Bits
import Data.Tuple (swap)
import Data.Word (Word8, Word16, Word32, Word64)

-- Generic parsing types --
type Parser a = L.ByteString
             -> Either (L.ByteString, String) a

type Peeler a = L.ByteString
             -> Either (L.ByteString, String) (L.ByteString, a)
type Expecter a = a
               -> L.ByteString
               -> Either (L.ByteString, String) L.ByteString

peeler :: String
       -> (L.ByteString -> Maybe (L.ByteString, a))
       -> Peeler a
peeler err f s = maybe (Left (s, err)) Right $ f s

expecter :: Eq a => Show a
         => Peeler a -> Expecter a
expecter p a s = do
  (s', a') <- p s
  if a' == a
    then Right s'
    else Left (s, "expected " ++ show a ++ " found " ++ show a')

parser :: Peeler a -> Parser a
parser p s = do
  (s', a) <- p s
  if L.length s' /= 0
    then Left (s', "expected no more bytes")
    else Right a

peelSome :: Peeler a -> Peeler [a]
peelSome p s = peelAny p s >>=
  \(s', as) -> if length as == 0
               then Left (s, "expected more than one!")
               else Right (s', as)

peelAny :: Peeler a -> Peeler [a]
peelAny p s =
  case p s of
    Left (_, _) -> Right (s, [])
    Right (s', a) -> msnd (a:) <$> (peelAny p s')

assertThat :: L.ByteString
           -> Bool
           -> Either (L.ByteString, String) ()
assertThat _ True  = Right ()
assertThat s False = Left (s, "assert was false")


-- Handy peelers and parsers --
peelNBytes :: Int -> Peeler L.ByteString
peelNBytes n = peeler ("expected " ++ show n ++ " bytes") $
  \s -> if L.length s >= n'
        then Just $ swap $ L.splitAt n' s
        else Nothing
  where n' = fromIntegral n

expectBytes :: Expecter L.ByteString
expectBytes bs = expecter (peelNBytes $ fromIntegral $ L.length bs) bs

mfst :: (a -> a') -> (a, b) -> (a', b)
mfst f (a, b) = (f a, b)
msnd :: (b -> b') -> (a, b) -> (a, b')
msnd f (a, b) = (a, f b)

peelWord8 :: Peeler Word8
peelWord8 s = msnd L.head <$> peelNBytes 1 s

peelWord16 :: Peeler Word16
peelWord16 s = msnd toWord16 <$> peelNBytes 2 s
  where
    toWord16 bs = fromIntegral (L.index bs 1)
                  + shiftL (fromIntegral (L.head bs)) 8

peelWord32 :: Peeler Word32
peelWord32 s = msnd toWord32 <$> peelNBytes 4 s
  where
    toWord32 bs = fromIntegral (L.index bs 3)
                  + shiftL (fromIntegral (L.index bs 2)) 8
                  + shiftL (fromIntegral (L.index bs 1)) 16
                  + shiftL (fromIntegral (L.index bs 0)) 24

peelWord64 :: Peeler Word64
peelWord64 s = msnd toWord64 <$> peelNBytes 4 s
  where
    toWord64 bs = fromIntegral (L.index bs 7)
                  + shiftL (fromIntegral (L.index bs 6)) 8
                  + shiftL (fromIntegral (L.index bs 5)) 16
                  + shiftL (fromIntegral (L.index bs 4)) 24
                  + shiftL (fromIntegral (L.index bs 3)) 32
                  + shiftL (fromIntegral (L.index bs 2)) 40
                  + shiftL (fromIntegral (L.index bs 1)) 48
                  + shiftL (fromIntegral (L.index bs 0)) 56


jpeg_parser.hs
import qualified Data.ByteString.Lazy.Char8 as L8
import qualified Data.ByteString.Lazy as L
import qualified Data.ByteString as B
import Data.Char (isSpace)
import Data.Word (Word8, Word16, Word32, Word64)
import Parse
import Numeric (showHex)
import Data.Either (either)

-- Jpeg
data Jpeg = Jpeg {
    jpegWidth :: Int
  , jpegHeight :: Int
  , jpegDQTable :: DQTable
  , jpegHuffmanTable :: HuffmanTable
  , jpegBytes :: B.ByteString
  } deriving (Eq)

instance Show Jpeg where
  show (Jpeg w h _ _) = "Jpeg " ++ show w ++ "x" ++ show h

parseJpeg :: Parser Jpeg
parseJpeg bytes = do
  s <- expectMarker soiMarker bytes
  (s', seg0) <- peelApp0Segment s
  (s', segs) <- peelAny peelAppnSegment s'
  (s', dqt) <- peelDqt s'
  Left (s', "not done implementing parseJpeg: " ++ show seg0)

debug :: L.ByteString -> Either String Jpeg
debug s = case parseJpeg s of
  Left (s', err) -> Left $
                    lengthMessage s' s ++ "\n"
                    ++ show (L.take 15 s') ++ "...\n"
                    ++ err
  Right j -> Right j
  where lengthMessage s' s =
          let read = fromIntegral $ L.length s - L.length s'
          in "read " ++ show read ++ " (0x" ++ showHex read "" ++ ") bytes. " ++ show (L.length s') ++ " to go"

doDebug :: L.ByteString -> IO ()
doDebug s = putStrLn $ either id show $ debug s

-- AppnSegments
data AppnSegment = AppnSegment {
    appnN :: Int
  , appnBytes :: B.ByteString
  }

peelAppnSegment :: Peeler AppnSegment
peelAppnSegment s = do
  (s', m) <- peelMarker s
  let mb = fromIntegral $ markerByte m
  _ <- assertThat s $ 0xe0 <= mb && mb <= 0xef
  (s', length) <- peelWord16 s'
  (s', bytes) <- peelNBytes (fromIntegral length - 2) s'
  Right $ (,) s' $ AppnSegment (mb - 0xe0) (L.toStrict bytes)


data App0Segment = App0Segment {
    app0Major :: Word8
  , app0Minor :: Word8
  , app0DensityUnits :: DensityUnits
  , app0XDensity :: Int
  , app0YDensity :: Int
  , app0ThumbWidth :: Int
  , app0ThumbHeight :: Int
  , app0ThumbBytes :: B.ByteString
  } deriving (Eq)
data DensityUnits = NoUnits | DotsPerInch | DotsPerCm
  deriving (Eq, Show)
densityUnitsFromInt 0 = Just NoUnits
densityUnitsFromInt 1 = Just DotsPerInch
densityUnitsFromInt 2 = Just DotsPerCm
densityUnitsFromInt _ = Nothing

instance Show App0Segment where
  show seg = "App0Segment"
          ++ " v" ++ show (app0Major seg) ++ "." ++ show (app0Minor seg)
          ++ " " ++ show (app0DensityUnits seg)
          ++ " Density:" ++ show (app0XDensity seg) ++ ":" ++ show (app0YDensity seg)
          ++ " Thumb:" ++ show (app0ThumbWidth seg) ++ "x" ++ show (app0ThumbHeight seg)

peelApp0Segment :: Peeler App0Segment
peelApp0Segment s = do
  (s', seg) <- peelAppnSegment s
  _ <- assertThat s $ appnN seg == 0
  let bs = L.fromStrict $ appnBytes seg
  bs' <- expectBytes (L8.pack "JFIF\0") bs
  bs' <- expectBytes (L.pack [1]) bs'
  (bs', minorRevision) <- peelWord8 bs'
  (bs'', du) <- peelWord8 bs'
  densityUnits <- maybe (Left (bs', "expected 0-2 for density units")) Right $ densityUnitsFromInt du
  (bs', xDensity) <- peelWord16 bs''
  (bs', yDensity) <- peelWord16 bs'
  (bs', thumbWidth) <- peelWord8 bs'
  (bs', thumbHeight) <- peelWord8 bs'
  let thumbLength = 3 * (fromIntegral thumbWidth) * (fromIntegral thumbHeight)
  (bs', thumbBytes) <- peelNBytes thumbLength bs'
  _ <- assertThat bs' $ L.length bs' == 0
  return $ (,) s' $
    App0Segment 1 minorRevision densityUnits (fromIntegral xDensity) (fromIntegral yDensity) (fromIntegral thumbWidth) (fromIntegral thumbHeight) (L.toStrict thumbBytes)


-- Discrete Quantization Table
data DQTable = DQTable {
    dqtLength :: Int
  , dqtBytes :: B.ByteString
  }

peelDqt :: Peeler DQTable
peelDqt s = do
  s' <- expectMarker dqtMarker s
  (s', length) <- peelWord16 s'
  (s', bytes) <- peelNBytes (fromIntegral length - 2) s'
  return $ (,) s' $
    DQTable (fromIntegral length) (L.toStrict bytes)

-- Marker
data Marker = Marker { markerByte :: Word8 }
  deriving (Eq, Show, Read)

isMarker :: L.ByteString -> Bool
isMarker s = L.length s == 2 && L.head s == 0xff

soiMarker    = Marker 0xd8
appnMarker n = Marker $ 0xe0 + n
app0Marker   = appnMarker 0
dqtMarker    = Marker 0xdb
sofnMarker n = Marker $ 0xc0 + n
dhtMarker    = Marker 0xc4
sosMarker    = Marker 0xda
eoiMarker    = Marker 0xd9

expectMarker :: Expecter Marker
expectMarker = expecter peelMarker

peelMarker :: Peeler Marker
peelMarker = peeler "expected marker" maybeGetMarker
  where
    maybeGetMarker s =
      let (mbs, s') = L.splitAt 2 s in
        if not $ isMarker mbs
        then Nothing
        else Just $ (s', Marker $ L.index mbs 1)

stripToMarker :: L.ByteString
              -> L.ByteString
              -> L.ByteString
stripToMarker m xs | xs == L.empty = L.empty
stripToMarker m xs =
  if L.take 2 ffOn == m
  then ffOn
  else stripToMarker m $ L.drop 1 ffOn
  where
    ffOn = stripToFf xs
    stripToFf = L.dropWhile $ \w -> w /= 0xff


-- Huffman Coding --

data HuffmanTable = HuffmanTable {} deriving (Eq)


Jpeg.hs
module Jpeg where

import qualified Data.ByteString.Lazy.Char8 as L8
import qualified Data.ByteString.Lazy as L
import qualified Data.ByteString as B
import Data.Word (Word8, Word16, Word32, Word64)
import Parse
import Numeric (showHex)


-- Jpeg
data Jpeg = Jpeg {
    jpegApp0 :: App0Segment
  , jpegAppns :: [AppnSegment]
  , jpegDQTable :: [DQTable]
  , jpegSof :: SOFSegment
  , jpegDHTable :: [DHTable]
  , jpegSos :: SOSSegment
  , jpegBytes :: B.ByteString
  } deriving (Eq)

instance Show Jpeg where
  show (Jpeg _ _ _ _ _ _ _) = "Jpeg"

parseJpeg :: Parser Jpeg
parseJpeg = parser peelJpeg

peelJpeg :: Peeler Jpeg
peelJpeg bytes = do
  s <- expectMarker soiMarker bytes
  (s', seg0) <- peelApp0Segment s
  (s', segs) <- peelAny peelAppnSegment s'
  (s', dqts) <- peelSome peelDqt s'
  (s', sof) <- peelSof s'
  (s', dht) <- peelSome peelDht s'
  (s', sos) <- peelSos s'
  (s', imageBytes) <- msnd L.toStrict <$> peelImageData s'
  s' <- expectMarker eoiMarker s'
  Right $ (,) s' $
    Jpeg seg0 segs dqts sof dht sos imageBytes

debug :: L.ByteString -> Either String Jpeg
debug s = case parseJpeg s of
  Left (s', err) -> Left $
                    lengthMessage s' s ++ "\n"
                    ++ "left: "
                    ++ (if L.length left < L.length s'
                         then show left ++ "..." else show left)
                    ++ "\n"
                    ++ err
                    where left = L.take 15 s'
  Right j -> Right j
  where lengthMessage s' s =
          let read = fromIntegral $ L.length s - L.length s'
          in "read: " ++ show read ++ " (0x" ++ showHex read "" ++ ") bytes. " ++ show (L.length s') ++ " to go"

doDebug :: String -> IO ()
doDebug fn = putStrLn . either id show . debug =<< L.readFile fn


-- AppnSegments
data AppnSegment = AppnSegment {
    appnN :: Int
  , appnBytes :: B.ByteString
  } deriving (Eq)

peelAppnSegment :: Peeler AppnSegment
peelAppnSegment s = do
  (s', m) <- peelMarker s
  let mb = fromIntegral $ markerByte m
  _ <- assertThat s $ 0xe0 <= mb && mb <= 0xef
  (s', bs) <- peelSegment m s
  return $ (,) s' $
    AppnSegment (mb - 0xe0) (L.toStrict bs)


data App0Segment = App0Segment {
    app0Major :: Word8
  , app0Minor :: Word8
  , app0DensityUnits :: DensityUnits
  , app0XDensity :: Int
  , app0YDensity :: Int
  , app0ThumbWidth :: Int
  , app0ThumbHeight :: Int
  , app0ThumbBytes :: B.ByteString
  } deriving (Eq)
data DensityUnits = NoUnits | DotsPerInch | DotsPerCm
  deriving (Eq, Show)
densityUnitsFromInt 0 = Just NoUnits
densityUnitsFromInt 1 = Just DotsPerInch
densityUnitsFromInt 2 = Just DotsPerCm
densityUnitsFromInt _ = Nothing

instance Show App0Segment where
  show seg = "App0Segment"
          ++ " v" ++ show (app0Major seg) ++ "." ++ show (app0Minor seg)
          ++ " " ++ show (app0DensityUnits seg)
          ++ " Density:" ++ show (app0XDensity seg) ++ ":" ++ show (app0YDensity seg)
          ++ " Thumb:" ++ show (app0ThumbWidth seg) ++ "x" ++ show (app0ThumbHeight seg)

peelApp0Segment :: Peeler App0Segment
peelApp0Segment s = do
  (s', seg) <- peelAppnSegment s
  _ <- assertThat s $ appnN seg == 0
  let bs = L.fromStrict $ appnBytes seg
  bs' <- expectBytes (L8.pack "JFIF\0") bs
  bs' <- expectBytes (L.pack [1]) bs'
  (bs', minorRevision) <- peelWord8 bs'
  (bs'', du) <- peelWord8 bs'
  densityUnits <-
    maybe (Left (bs', "expected 0-2 for density units"))
          Right
          (densityUnitsFromInt du)
  (bs', xDensity) <- peelWord16 bs''
  (bs', yDensity) <- peelWord16 bs'
  (bs', thumbWidth) <- peelWord8 bs'
  (bs', thumbHeight) <- peelWord8 bs'
  let thumbLength = 3 * (fromIntegral thumbWidth) * (fromIntegral thumbHeight)
  (bs', thumbBytes) <- peelNBytes thumbLength bs'
  _ <- assertThat bs' $ L.length bs' == 0
  return $ (,) s' $
    App0Segment 1 minorRevision densityUnits (fromIntegral xDensity) (fromIntegral yDensity) (fromIntegral thumbWidth) (fromIntegral thumbHeight) (L.toStrict thumbBytes)


-- Discrete Quantization Table
data DQTable = DQTable {
    dqtBytes :: B.ByteString
  } deriving (Eq)

peelDqt :: Peeler DQTable
peelDqt s = do
  (s', bs) <- peelSegment dqtMarker s
  return $ (,) s' $
    DQTable $ L.toStrict bs


-- Start of Frame Segment
data SOFSegment = SOFSegment {
    sofBytes :: B.ByteString
  } deriving (Eq)

peelSof :: Peeler SOFSegment
peelSof s = do
  (s', bs) <- peelSegment sofMarker s
  return $ (,) s' $
    SOFSegment $ L.toStrict bs


-- Huffman Table
data DHTable = DHTable {
    dhtBytes :: B.ByteString
  } deriving (Eq)

peelDht :: Peeler DHTable
peelDht s = do
  (s', bs) <- peelSegment dhtMarker s
  return $ (,) s' $
    DHTable $ L.toStrict bs


-- Start of Scan (actual image)
data SOSSegment = SOSSegment {
    sosBytes :: B.ByteString
  } deriving (Eq)

peelSos :: Peeler SOSSegment
peelSos s = do
  (s', bs) <- peelSegment sosMarker s
  return $ (,) s' $
    SOSSegment $ L.toStrict bs


-- Image data
peelImageData :: Peeler L.ByteString
peelImageData s = Right $ msnd (L.concat . reverse) $ helper [] s
  where helper l s =
          let (hd, tl) = L.break (==0xff) s
          in if L.index tl 1 == 0x00
             then helper ((L.pack [0xff]):hd:l) $ L.drop 2 tl
             else (tl, hd:l)


-- Marker
data Marker = Marker { markerByte :: Word8 }
  deriving (Eq, Read)
instance Show Marker where
  show (Marker b) = "Marker 0x" ++ showHex b ""

isMarker :: L.ByteString -> Bool
isMarker s = L.length s == 2 && L.head s == 0xff

soiMarker    = Marker 0xd8
appnMarker n = Marker $ 0xe0 + n
app0Marker   = appnMarker 0
dqtMarker    = Marker 0xdb
sofMarker    = Marker $ 0xc0
dhtMarker    = Marker 0xc4
sosMarker    = Marker 0xda
eoiMarker    = Marker 0xd9

expectMarker :: Expecter Marker
expectMarker = expecter peelMarker

peelMarker :: Peeler Marker
peelMarker = peeler "expected marker" maybeGetMarker
  where
    maybeGetMarker s =
      let (mbs, s') = L.splitAt 2 s in
        if not $ isMarker mbs
        then Nothing
        else Just $ (s', Marker $ L.index mbs 1)

peelSegment :: Marker -> Peeler L.ByteString
peelSegment m s = do
  s' <- expectMarker m s
  (s', l) <- peelWord16 s'
  (s', bs) <- peelNBytes (fromIntegral l - 2) s'
  return (s', bs)
